File: src/native/modrm.c
#include <assert.h>
#include <math.h>
#include <stdbool.h>
#include <stdint.h>

#include "const.h"
#include "cpu.h"
#include "global_pointers.h"
#include "modrm.h"

//static int32_t resolve_sib_(bool mod);
//static int32_t resolve_modrm32_(int32_t modrm_byte);
static int32_t resolve_sib(bool mod);

#define ds get_seg_prefix_ds
#define ss get_seg_prefix_ss

#define MODRM_ENTRY(n, offset)\
    case (n) | 0 << 3:\
    case (n) | 1 << 3:\
    case (n) | 2 << 3:\
    case (n) | 3 << 3:\
    case (n) | 4 << 3:\
    case (n) | 5 << 3:\
    case (n) | 6 << 3:\
    case (n) | 7 << 3:\
        return offset;

#define MODRM_ENTRY16(row, seg, value)\
    MODRM_ENTRY(0x00 | (row), seg(((value) & 0xFFFF)))\
    MODRM_ENTRY(0x40 | (row), seg(((value) + read_imm8s() & 0xFFFF)))\
    MODRM_ENTRY(0x80 | (row), seg(((value) + read_imm16() & 0xFFFF)))\

int32_t resolve_modrm16(int32_t modrm_byte)
{
    switch(modrm_byte)
    {
        MODRM_ENTRY16(0, ds, reg16[BX] + reg16[SI])
        MODRM_ENTRY16(1, ds, reg16[BX] + reg16[DI])
        MODRM_ENTRY16(2, ss, reg16[BP] + reg16[SI])
        MODRM_ENTRY16(3, ss, reg16[BP] + reg16[DI])
        MODRM_ENTRY16(4, ds, reg16[SI])
        MODRM_ENTRY16(5, ds, reg16[DI])

        // special case
        MODRM_ENTRY(0x00 | 6, ds(read_imm16()))
        MODRM_ENTRY(0x40 | 6, ss(reg16[BP] + read_imm8s() & 0xFFFF))
        MODRM_ENTRY(0x80 | 6, ss(reg16[BP] + read_imm16() & 0xFFFF))

        MODRM_ENTRY16(7, ds, reg16[BX])

        default:
            assert(false);
    }

    return 0;
}

#undef MODRM_ENTRY16

#define MODRM_ENTRY32(row, seg, value)\
    MODRM_ENTRY(0x00 | (row), seg((value)))\
    MODRM_ENTRY(0x40 | (row), seg((value) + read_imm8s()))\
    MODRM_ENTRY(0x80 | (row), seg((value) + read_imm32s()))\

int32_t resolve_modrm32(int32_t modrm_byte)
{
    switch(modrm_byte)
    {
        MODRM_ENTRY32(0, ds, reg32s[EAX])
        MODRM_ENTRY32(1, ds, reg32s[ECX])
        MODRM_ENTRY32(2, ds, reg32s[EDX])
        MODRM_ENTRY32(3, ds, reg32s[EBX])

        // special cases
        MODRM_ENTRY(0x00 | 4, resolve_sib(false))
        MODRM_ENTRY(0x40 | 4, resolve_sib(true) + read_imm8s())
        MODRM_ENTRY(0x80 | 4, resolve_sib(true) + read_imm32s())
        MODRM_ENTRY(0x00 | 5, ds(read_imm32s()))
        MODRM_ENTRY(0x40 | 5, ss(reg32s[EBP] + read_imm8s()))
        MODRM_ENTRY(0x80 | 5, ss(reg32s[EBP] + read_imm32s()))

        MODRM_ENTRY32(6, ds, reg32s[ESI])
        MODRM_ENTRY32(7, ds, reg32s[EDI])

        default:
            assert(false);
    }

    return 0;
}

#undef MODRM_ENTRY32
#undef MODRM_ENTRY


#define SIB_ENTRY_LEVEL3(n, offset)\
    case n: return offset;

#define SIB_ENTRY_LEVEL2(n, offset)\
    SIB_ENTRY_LEVEL3((n) | 0, ds((offset) + reg32s[EAX]))\
    SIB_ENTRY_LEVEL3((n) | 1, ds((offset) + reg32s[ECX]))\
    SIB_ENTRY_LEVEL3((n) | 2, ds((offset) + reg32s[EDX]))\
    SIB_ENTRY_LEVEL3((n) | 3, ds((offset) + reg32s[EBX]))\
    SIB_ENTRY_LEVEL3((n) | 4, ss((offset) + reg32s[ESP]))\
    SIB_ENTRY_LEVEL3((n) | 5, (mod ? ss((offset) + reg32s[EBP]) : ds((offset) + read_imm32s())))\
    SIB_ENTRY_LEVEL3((n) | 6, ds((offset) + reg32s[ESI]))\
    SIB_ENTRY_LEVEL3((n) | 7, ds((offset) + reg32s[EDI]))

#define SIB_ENTRY_LEVEL1(n, reg1)\
    SIB_ENTRY_LEVEL2(0x00 | (n) << 3, (reg1))\
    SIB_ENTRY_LEVEL2(0x40 | (n) << 3, (reg1) << 1)\
    SIB_ENTRY_LEVEL2(0x80 | (n) << 3, (reg1) << 2)\
    SIB_ENTRY_LEVEL2(0xC0 | (n) << 3, (reg1) << 3)

#if 0
static inline int32_t resolve_sib_(bool mod)
{
    switch(read_imm8())
    {
        SIB_ENTRY_LEVEL1(0, reg32s[EAX]);
        SIB_ENTRY_LEVEL1(1, reg32s[ECX]);
        SIB_ENTRY_LEVEL1(2, reg32s[EDX]);
        SIB_ENTRY_LEVEL1(3, reg32s[EBX]);
        SIB_ENTRY_LEVEL1(4, 0          );
        SIB_ENTRY_LEVEL1(5, reg32s[EBP]);
        SIB_ENTRY_LEVEL1(6, reg32s[ESI]);
        SIB_ENTRY_LEVEL1(7, reg32s[EDI]);

        default:
            assert(false);
    }

    return 0;
}
#endif

#undef SIB_ENTRY_LEVEL3
#undef SIB_ENTRY_LEVEL2
#undef SIB_ENTRY_LEVEL1

#undef ds
#undef ss


static int32_t resolve_sib(bool mod)
{
    uint8_t sib_byte = read_imm8();
    uint8_t r = sib_byte & 7;
    uint8_t m = sib_byte >> 3 & 7;

    int32_t base;
    int32_t seg;

    if(r == 4)
    {
        base = reg32s[ESP];
        seg = SS;
    }
    else if(r == 5)
    {
        if(mod)
        {
            base = reg32s[EBP];
            seg = SS;
        }
        else
        {
            base = read_imm32s();
            seg = DS;
        }
    }
    else
    {
        base = reg32s[r];
        seg = DS;
    }

    int32_t offset;
    if(m == 4)
    {
        offset = 0;
    }
    else
    {
        uint8_t s = sib_byte >> 6 & 3;
        offset = reg32s[m] << s;
    }

    return get_seg_prefix(seg) + base + offset;
}

#if 0
static inline int32_t resolve_modrm32_(int32_t modrm_byte)
{
    uint8_t r = modrm_byte & 7;
    assert(modrm_byte < 0xC0);

    if(r == 4)
    {
        if(modrm_byte < 0x40)
        {
            return resolve_sib(false);
        }
        else
        {
            return resolve_sib(true) + (modrm_byte < 0x80 ? read_imm8s() : read_imm32s());
        }
    }
    else if(r == 5)
    {
        if(modrm_byte < 0x40)
        {
            return get_seg_prefix_ds(read_imm32s());
        }
        else
        {
            return get_seg_prefix_ss(reg32s[EBP] + (modrm_byte < 0x80 ? read_imm8s() : read_imm32s()));
        }
    }
    else
    {
        if(modrm_byte < 0x40)
        {
            return get_seg_prefix_ds(reg32s[r]);
        }
        else
        {
            return get_seg_prefix_ds(reg32s[r] + (modrm_byte < 0x80 ? read_imm8s() : read_imm32s()));
        }
    }
}
#endif

File: src/rust/cpu2/modrm.rs
use cpu2::cpu::*;
use cpu2::global_pointers::*;
use paging::OrPageFault;

pub unsafe fn resolve_modrm16(modrm_byte: i32) -> OrPageFault<i32> {
    Ok(match modrm_byte {
        0 | 8 | 16 | 24 | 32 | 40 | 48 | 56 => get_seg_prefix_ds(
            *reg16.offset(BX as isize) as i32 + *reg16.offset(SI as isize) as i32 & 0xFFFF,
        )?,
        64 | 72 | 80 | 88 | 96 | 104 | 112 | 120 => get_seg_prefix_ds(
            *reg16.offset(BX as isize) as i32 + *reg16.offset(SI as isize) as i32 + read_imm8s()?
                & 0xFFFF,
        )?,
        128 | 136 | 144 | 152 | 160 | 168 | 176 | 184 => get_seg_prefix_ds(
            *reg16.offset(BX as isize) as i32 + *reg16.offset(SI as isize) as i32 + read_imm16()?
                & 0xFFFF,
        )?,
        1 | 9 | 17 | 25 | 33 | 41 | 49 | 57 => get_seg_prefix_ds(
            *reg16.offset(BX as isize) as i32 + *reg16.offset(DI as isize) as i32 & 0xFFFF,
        )?,
        65 | 73 | 81 | 89 | 97 | 105 | 113 | 121 => get_seg_prefix_ds(
            *reg16.offset(BX as isize) as i32 + *reg16.offset(DI as isize) as i32 + read_imm8s()?
                & 0xFFFF,
        )?,
        129 | 137 | 145 | 153 | 161 | 169 | 177 | 185 => get_seg_prefix_ds(
            *reg16.offset(BX as isize) as i32 + *reg16.offset(DI as isize) as i32 + read_imm16()?
                & 0xFFFF,
        )?,
        2 | 10 | 18 | 26 | 34 | 42 | 50 | 58 => get_seg_prefix_ss(
            *reg16.offset(BP as isize) as i32 + *reg16.offset(SI as isize) as i32 & 0xFFFF,
        )?,
        66 | 74 | 82 | 90 | 98 | 106 | 114 | 122 => get_seg_prefix_ss(
            *reg16.offset(BP as isize) as i32 + *reg16.offset(SI as isize) as i32 + read_imm8s()?
                & 0xFFFF,
        )?,
        130 | 138 | 146 | 154 | 162 | 170 | 178 | 186 => get_seg_prefix_ss(
            *reg16.offset(BP as isize) as i32 + *reg16.offset(SI as isize) as i32 + read_imm16()?
                & 0xFFFF,
        )?,
        3 | 11 | 19 | 27 | 35 | 43 | 51 | 59 => get_seg_prefix_ss(
            *reg16.offset(BP as isize) as i32 + *reg16.offset(DI as isize) as i32 & 0xFFFF,
        )?,
        67 | 75 | 83 | 91 | 99 | 107 | 115 | 123 => get_seg_prefix_ss(
            *reg16.offset(BP as isize) as i32 + *reg16.offset(DI as isize) as i32 + read_imm8s()?
                & 0xFFFF,
        )?,
        131 | 139 | 147 | 155 | 163 | 171 | 179 | 187 => get_seg_prefix_ss(
            *reg16.offset(BP as isize) as i32 + *reg16.offset(DI as isize) as i32 + read_imm16()?
                & 0xFFFF,
        )?,
        4 | 12 | 20 | 28 | 36 | 44 | 52 | 60 => {
            get_seg_prefix_ds(*reg16.offset(SI as isize) as i32 & 0xFFFF)?
        },
        68 | 76 | 84 | 92 | 100 | 108 | 116 | 124 => {
            get_seg_prefix_ds(*reg16.offset(SI as isize) as i32 + read_imm8s()? & 0xFFFF)?
        },
        132 | 140 | 148 | 156 | 164 | 172 | 180 | 188 => {
            get_seg_prefix_ds(*reg16.offset(SI as isize) as i32 + read_imm16()? & 0xFFFF)?
        },
        5 | 13 | 21 | 29 | 37 | 45 | 53 | 61 => {
            get_seg_prefix_ds(*reg16.offset(DI as isize) as i32 & 0xFFFF)?
        },
        69 | 77 | 85 | 93 | 101 | 109 | 117 | 125 => {
            get_seg_prefix_ds(*reg16.offset(DI as isize) as i32 + read_imm8s()? & 0xFFFF)?
        },
        133 | 141 | 149 | 157 | 165 | 173 | 181 | 189 => {
            get_seg_prefix_ds(*reg16.offset(DI as isize) as i32 + read_imm16()? & 0xFFFF)?
        },
        6 | 14 | 22 | 30 | 38 | 46 | 54 | 62 => get_seg_prefix_ds(read_imm16()?)?,
        70 | 78 | 86 | 94 | 102 | 110 | 118 | 126 => {
            get_seg_prefix_ss(*reg16.offset(BP as isize) as i32 + read_imm8s()? & 0xFFFF)?
        },
        134 | 142 | 150 | 158 | 166 | 174 | 182 | 190 => {
            get_seg_prefix_ss(*reg16.offset(BP as isize) as i32 + read_imm16()? & 0xFFFF)?
        },
        7 | 15 | 23 | 31 | 39 | 47 | 55 | 63 => {
            get_seg_prefix_ds(*reg16.offset(BX as isize) as i32 & 0xFFFF)?
        },
        71 | 79 | 87 | 95 | 103 | 111 | 119 | 127 => {
            get_seg_prefix_ds(*reg16.offset(BX as isize) as i32 + read_imm8s()? & 0xFFFF)?
        },
        135 | 143 | 151 | 159 | 167 | 175 | 183 | 191 => {
            get_seg_prefix_ds(*reg16.offset(BX as isize) as i32 + read_imm16()? & 0xFFFF)?
        },
        _ => {
            dbg_assert!(false);
            0
        },
    })
}

pub unsafe fn resolve_modrm32_(modrm_byte: i32) -> OrPageFault<i32> {
    let r = (modrm_byte & 7) as u8;
    dbg_assert!(modrm_byte < 192);
    Ok(if r as i32 == 4 {
        if modrm_byte < 64 {
            resolve_sib(false)?
        }
        else {
            resolve_sib(true)? + if modrm_byte < 128 { read_imm8s()? } else { read_imm32s()? }
        }
    }
    else if r as i32 == 5 {
        if modrm_byte < 64 {
            get_seg_prefix_ds(read_imm32s()?)?
        }
        else {
            get_seg_prefix_ss(
                *reg32.offset(EBP as isize)
                    + if modrm_byte < 128 { read_imm8s()? } else { read_imm32s()? },
            )?
        }
    }
    else if modrm_byte < 64 {
        get_seg_prefix_ds(*reg32.offset(r as isize))?
    }
    else {
        get_seg_prefix_ds(
            *reg32.offset(r as isize)
                + if modrm_byte < 128 { read_imm8s()? } else { read_imm32s()? },
        )?
    })
}
unsafe fn resolve_sib(mod_0: bool) -> OrPageFault<i32> {
    let s;
    let sib_byte = read_imm8()? as u8;
    let r = (sib_byte as i32 & 7) as u8;
    let m = (sib_byte as i32 >> 3 & 7) as u8;
    let base;
    let seg;
    if r as i32 == 4 {
        base = *reg32.offset(ESP as isize);
        seg = SS
    }
    else if r as i32 == 5 {
        if mod_0 {
            base = *reg32.offset(EBP as isize);
            seg = SS
        }
        else {
            base = read_imm32s()?;
            seg = DS
        }
    }
    else {
        base = *reg32.offset(r as isize);
        seg = DS
    }
    let offset;
    if m as i32 == 4 {
        offset = 0
    }
    else {
        s = (sib_byte as i32 >> 6 & 3) as u8;
        offset = *reg32.offset(m as isize) << s as i32
    }
    Ok(get_seg_prefix(seg)? + base + offset)
}

pub unsafe fn resolve_modrm32(modrm_byte: i32) -> OrPageFault<i32> {
    Ok(match modrm_byte {
        0 | 8 | 16 | 24 | 32 | 40 | 48 | 56 => get_seg_prefix_ds(*reg32.offset(EAX as isize))?,
        64 | 72 | 80 | 88 | 96 | 104 | 112 | 120 => {
            get_seg_prefix_ds(*reg32.offset(EAX as isize) + read_imm8s()?)?
        },
        128 | 136 | 144 | 152 | 160 | 168 | 176 | 184 => {
            get_seg_prefix_ds(*reg32.offset(EAX as isize) + read_imm32s()?)?
        },
        1 | 9 | 17 | 25 | 33 | 41 | 49 | 57 => get_seg_prefix_ds(*reg32.offset(ECX as isize))?,
        65 | 73 | 81 | 89 | 97 | 105 | 113 | 121 => {
            get_seg_prefix_ds(*reg32.offset(ECX as isize) + read_imm8s()?)?
        },
        129 | 137 | 145 | 153 | 161 | 169 | 177 | 185 => {
            get_seg_prefix_ds(*reg32.offset(ECX as isize) + read_imm32s()?)?
        },
        2 | 10 | 18 | 26 | 34 | 42 | 50 | 58 => get_seg_prefix_ds(*reg32.offset(EDX as isize))?,
        66 | 74 | 82 | 90 | 98 | 106 | 114 | 122 => {
            get_seg_prefix_ds(*reg32.offset(EDX as isize) + read_imm8s()?)?
        },
        130 | 138 | 146 | 154 | 162 | 170 | 178 | 186 => {
            get_seg_prefix_ds(*reg32.offset(EDX as isize) + read_imm32s()?)?
        },
        3 | 11 | 19 | 27 | 35 | 43 | 51 | 59 => get_seg_prefix_ds(*reg32.offset(EBX as isize))?,
        67 | 75 | 83 | 91 | 99 | 107 | 115 | 123 => {
            get_seg_prefix_ds(*reg32.offset(EBX as isize) + read_imm8s()?)?
        },
        131 | 139 | 147 | 155 | 163 | 171 | 179 | 187 => {
            get_seg_prefix_ds(*reg32.offset(EBX as isize) + read_imm32s()?)?
        },
        4 | 12 | 20 | 28 | 36 | 44 | 52 | 60 => resolve_sib(false)?,
        68 | 76 | 84 | 92 | 100 | 108 | 116 | 124 => resolve_sib(true)? + read_imm8s()?,
        132 | 140 | 148 | 156 | 164 | 172 | 180 | 188 => resolve_sib(true)? + read_imm32s()?,
        5 | 13 | 21 | 29 | 37 | 45 | 53 | 61 => get_seg_prefix_ds(read_imm32s()?)?,
        69 | 77 | 85 | 93 | 101 | 109 | 117 | 125 => {
            get_seg_prefix_ss(*reg32.offset(EBP as isize) + read_imm8s()?)?
        },
        133 | 141 | 149 | 157 | 165 | 173 | 181 | 189 => {
            get_seg_prefix_ss(*reg32.offset(EBP as isize) + read_imm32s()?)?
        },
        6 | 14 | 22 | 30 | 38 | 46 | 54 | 62 => get_seg_prefix_ds(*reg32.offset(ESI as isize))?,
        70 | 78 | 86 | 94 | 102 | 110 | 118 | 126 => {
            get_seg_prefix_ds(*reg32.offset(ESI as isize) + read_imm8s()?)?
        },
        134 | 142 | 150 | 158 | 166 | 174 | 182 | 190 => {
            get_seg_prefix_ds(*reg32.offset(ESI as isize) + read_imm32s()?)?
        },
        7 | 15 | 23 | 31 | 39 | 47 | 55 | 63 => get_seg_prefix_ds(*reg32.offset(EDI as isize))?,
        71 | 79 | 87 | 95 | 103 | 111 | 119 | 127 => {
            get_seg_prefix_ds(*reg32.offset(EDI as isize) + read_imm8s()?)?
        },
        135 | 143 | 151 | 159 | 167 | 175 | 183 | 191 => {
            get_seg_prefix_ds(*reg32.offset(EDI as isize) + read_imm32s()?)?
        },
        _ => {
            dbg_assert!(false);
            0
        },
    })
}
